py/builtinimport.c

```diff
@@ -133,6 +132,12 @@ STATIC void do_load(mp_obj_t module_obj, vstr_t *file) {
     mp_parse_compile_execute(lex, MP_PARSE_FILE_INPUT, mod_globals, mod_globals);
 }
 
+STATIC void do_load(mp_obj_t module_obj, vstr_t *file) {
+    // create the lexer
+    mp_lexer_t *lex = mp_lexer_new_from_file(vstr_str(file));
+    do_load_from_lexer(module_obj, lex, vstr_str(file));
```

Comment:
Yes, it's pretty tail-recursive, and I confirmed that when compiled with optimization on x64, the function ends with jump to do_load_from_lexer.